src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs

```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
                     }
 
                     // Process the attribute.
-                    string stringValue = null;
-                    if (assertion[0] is byte[] byteArray)
+                    byte[] byteArray;
+                    if (assertion[0] is string str)
                     {
-                        if (byteArray != null && byteArray.Length != 0)
-                        {
-                            berValuePtr = new BerVal
-                            {
-                                bv_len = byteArray.Length,
-                                bv_val = Marshal.AllocHGlobal(byteArray.Length)
-                            };
-                            Marshal.Copy(byteArray, 0, berValuePtr.bv_val, byteArray.Length);
-                        }
+                        var encoder = new UTF8Encoding();
+                        byteArray = encoder.GetBytes(str);
+                    }
+                    else if (assertion[0] is Uri uri)
+                    {
+                        var encoder = new UTF8Encoding();
+                        byteArray = encoder.GetBytes(uri.ToString());
+                    }
+                    else if (assertion[0] is byte[] bytes)
+                    {
+                        byteArray = bytes;
```

Comment:
What happens if the bytes were not utf8 encoded?

Reply:
CompareRequest operation returns ResultCode.CompareFalse. I added a test case for this.

Reply:
Right, but I guess my question is: should we support that? Or is it expected that bytes will always be UTF8 encoded? I quickly looked at our [docs](https://docs.microsoft.com/en-us/dotnet/api/system.directoryservices.protocols.comparerequest.-ctor?view=dotnet-plat-ext-6.0#system-directoryservices-protocols-comparerequest-ctor(system-string-system-string-system-byte())) and we don't seem to mention anywhere that the bytes need to be UTF8 encoded.

Reply:
I don't want to block your contribution on that, if we think UTF8 is what should be expected then we can just update our docs to point that out, and perhaps consider in the future add another constructor to CompareRequest that takes in bytes[] and Encoding.

Reply:
String values are always utf-8 encoded on the wire (per [ldap spec](https://datatracker.ietf.org/doc/html/rfc4511#section-4.1.2)). The byte[] value is as-is on the wire.

Reply:
Sounds good, we can update the docs then to call this out, but we can do that separately from your PR